### ISSUE TITLE
Can we use Elm?

### ISSUE THREAD

**Issue by nelsonic** [dwyl/technology-stack#27]
My plan for tomorrow morning is to _attempt_ to answer the question:
> # _Can_ we use `elm-lang` for DWYL projects/products? What are the ***costs/benefits***? 

http://elm-lang.org/
https://medium.com/wraptime-io/after-react-elm-152bd6559cb1
http://elm-lang.org/blog/how-to-use-elm-at-work

`CC`: @SimonLab @nogainbar @sofiapoh @des-des @krosti @jrans @eliascodes @shouston3  @rjmk @sohilpandya @mk4111 @jackcarlisle @harrygfox @naazy & `anyone` else interested ...

> relates to: https://github.com/dwyl/technology-stack/issues/18

**Comment by shouston3** [comment]
@andrewMacmurray @ivanmauricio I think you guys might be interested in seeing this

**Comment by andrewMacmurray** [comment]
thanks @shouston3 would definitely be interested to hear what people have to say on this!

**Comment by nelsonic** [comment]
@andrewMacmurray / @ivanmauricio, apologies, I did not _mean_ to "_exclude_" anyone... 
I usually try _not_ to "_spam_" people with notifications. But if you gent are interested, please vote: https://twitter.com/nelsonic/status/795164336834617344 (_thanks!_)

And if you don't mind helping to "_spread the word_" please `re-tweet` ♻️ ❤️

**Comment by sohilpandya** [comment]
@nelsonic @andrewMacmurray @ivanmauricio I would love to hear **why?** Isn't it another thing to learn on top? 

Have a look at this http://stateofjs.com/2016/flavors/

I thought there was a reason we were sticking with the plain old vanilla JS. Although I'm open to hear more

**Comment by andrewMacmurray** [comment]
@nelsonic no worries. @ivanmauricio and I had a play around with `elm` yesterday and loved ❤️  it. We're gonna build a web app that lets people keep track of how much they've contributed back to FAC using `elm` for the UI.

**Comment by sohilpandya** [comment]
@andrewMacmurray that's great to hear. 

A few of us (including @iteles & @nogainbar) were thinking of making that a week 9/10 16 man project for the current cohort. But if you guys are already working on it then we'll get them to work on another one of the tools that'll make certain other things easier at FAC. 😻

**Comment by andrewMacmurray** [comment]
@sohilpandya ah I didn't realise that was a project for FAC9. We could come up with something different, we just want an excuse to get stuck into using it and thought it would be nice to make something that would be useful for FAC.

**Comment by andrewMacmurray** [comment]
@sohilpandya as far as why we were interested using it, we're functional programming nerds 🤓 Lol but there are some pretty compelling reasons to use it beyond that. This vid is worth a watch

https://youtu.be/R2FtMbb-nLs

**Comment by sohilpandya** [comment]
@andrewMacmurray I think you guys should go ahead with this, We are just in the very initial stages and pivoting is a big possibility for us 😳

**Comment by des-des** [comment]
https://github.com/elm-lang/virtual-dom/issues/29

**Comment by iteles** [comment]
@andrewMacmurray @ivanmauricio @sohilpandya Whilst this isn't the place to discuss the FAC stuff, I would mention that if we're building something that's useful for FAC, my main concern is allowing it to be easily extendable by future cohorts/volunteers (essentially [this](https://github.com/dwyl/learn-elm/issues/2#issuecomment-258654412) - other issues in that repo well worth a read).

**Comment by andrewMacmurray** [comment]
@iteles I totally understand that, but what about its utility as a potential learning resource? I'm happy to discuss this elsewhere but not sure where to!

**Comment by nelsonic** [comment]
@andrewMacmurray discussing building a _useful_ app with `elm-lang` is _perfect_ candidate for this thread! I _deliberately_ left the question as _open_ as possible to _encourage/invite discussion_! 🎉 

The _last_ thing I want to do is "_shut down_" an open discussion because it's "_off topic_" (_the standard reason/excuse why questions are "closed" on StackOverflow_) ... I'm _encouraged_ by the fact that **38 people** have ***voted*** on the [**Twitter Poll!!**](https://twitter.com/nelsonic/status/795164336834617344) (_please help get **everyone** to vote so we have as much representation as possible!_) even though there are few people who "don't like" or "_not interested_"... I'm trying to remain "neutral" to not "_influence_" the Vote while the poll is still open. 😉 

> our objective at this stage is to [_`"first seek to understand"`_](https://www.youtube.com/watch?v=TpNbW9MAarQ) the sentiment. 🙊 
> I want to be **100% Clear** that I would ***never*** "_force_" anyone to learn something they were not at least a _little_ "_curious_" about learning ... so we _need_ to _gauge_ how people _feel_! 

To be _clear_: in terms of having a "stack" that is _well-defined_ and _unambiguous_, I'm _way_ more interested in going from ES6 _straight_ to Elm in "Week 7" without wasting _time_ on the _noise_ that is (_let's be honest_) the Facebook-(Franken)-Stack ... The _reason_ I haven't (_personally_) "_embraced_" elm in the _past_ is captured here: https://github.com/dwyl/learn-elm/issues/2

> but, the fact that elm (_currently_) has `Zero` support for server-side-rendering (_and we still think progressive enhancement matters!!_) is not going to be an issue for _much_ longer which _one_ of the reasons I opened this thread...! Hopefully that will change _soon_ as noted by @des-des [above](https://github.com/dwyl/technology-stack/issues/27#issuecomment-258686077) and indeed by @evancz himself recently at [`elm-conf`](https://youtu.be/DSjbTC-hvqQ?t=32m22s) so I'm _excited_ that the only remaining "_barrier_" to me leaving the 💩  behind is _almost_ ready. 

Until the Poll closes tho we need to _contain_ our _excitement_ and just encourage people to **Vote**!

![image](https://cloud.githubusercontent.com/assets/194400/20054648/cc22beae-a4d5-11e6-9f4e-40c1350bba12.png)

**Comment by andrewMacmurray** [comment]
@nelsonic @iteles I'd definitely prefer to build something in Elm that's going to be used by people but irregardless of whether it's something that's directly useful to FAC (or anyone), I feel there may be a special case for learning here.

I'm aware that Elm is "one more thing to learn" but compare that with learning something like React - where learning `React` comes with a huge ecosystem of things to learn like `webpack`, `babel`, `es6`, `jsx`, `flow`, `redux`, `immutable`. Elm's design seems to incorporate a lot of the benefits of these in a cohesive package (`redux` was even inspired by `the elm architecture`).

At first glance, Elm reminds me a little bit of `Hapi`, it's a "batteries included" approach with a clear design philosophy based on some very sound principles, and not to mention a little bit "left of field"!

> on a side note - If we're talking about building software that's easy for people to extend and inexpensive for clients to hire devs to do this, Hapi (which we embrace fully at dwyl and fac) is not exactly a standard choice. Whilst it's arguably easier to pick up than something like elm it still covers a large surface area of plugins and apis.

But I feel the big opportunity here is to embrace functional programming. Functional Programming is **HARD**, but that "one more thing" that you learn is more foundational than any of the JavaScript that we learn. You can apply widely it in many other programming languages and it's a very important trend in modern programming. It's knowledge that won't be thrown away any time soon.

As far as using it for the DWYL stack, I'm not nearly experienced enough to say (there's a lot more to consider than just learning). But, my gut feeling is I think it would be a missed opportunity if future fac people or volunteers weren't encouraged to get stuck into an elm project, at least if they were a bit curious about it.

(Sorry if this is a bit off topic but really wanted to share my thoughts on this.)

**Comment by rjmk** [comment]
Great comments in this thread (and in https://github.com/dwyl/learn-elm/issues/2). I think all of the concerns carry some heft.

I think Elm is in a good place in terms of maturity, but some of the complaints around that do seem pretty forceful. First, server-side rendering. No reason why that should continue to be an issue, but I also couldn't predict when it will stop being one. Second, handing Elm code over to clients. If I were hiring a bunch of devs to build a product, I would not hesitate in choosing a "weird" language and investing in the hires. But handing over Elm code to (non-technical) clients does leave them with a logistical problem.

I think the only complaint I don't think is telling is that it's another thing to learn. The Martin Fowler article Nelson linked to made a pretty good case that learning a language vs any other tool is not particularly different.

Also, consider the amount of knowledge and learning you have to repeatedly acquire about your own codebase whenever something unexpectedly breaks in JavaScript. You have to learn so much about how the code you and your team wrote weeks ago just to make small changes, and the knowledge disappears. Yes, you can try and follow "best practices" in your code, but that's of limited use. It can be much less effort to learn a new language and a new way of working than just trying to learn that new way of working in a language that doesn't support it very naturally.

In my experience, enough competence with a language to contribute to a codebase is one of the easiest things a dev will be asked to learn (apart from the first time).

**Comment by nelsonic** [comment]
@andrewMacmurray, agreed. don't [_`apologize`_](https://www.youtube.com/watch?v=ZSM3w1v-A_Y) when what you are saying/writing is _all_ relevant! 👍 

@rjmk ***`Yes`***, I _agree_ that delivering `elm-lang` code to clients is a _tricky_ one especially for an NGO or Local Council/GOV ... through no fault of their own they are (_technology_) "_risk-averse_" and don't _care_ about ["performance benchmarks"](http://elm-lang.org/blog/blazing-fast-html) or [_blog posts_](https://ohanhi.github.io/master-elm-1-why-elm.html) about "tech stack" ...
**`And`** we have to be _cautious_ on client's behalf (_until we have a comprehensive `learn-elm` tutorial we can share with people to help get them up-and-running fast..._) to _re-assure_ everyone.

I can _debate `both` "sides"_ of this conversation _convincingly_ with _plenty_ of supporting material. 💬 
**`And`** I _intend_ to take this discussion to it's _logical_ conclusion (_don't worry, you will_ 😍  _the outcome!_) the [**`#NextAction`**](https://github.com/dwyl/learn-elm/issues/4) is to produce a _comprehensive_ "_from-scratch_" tutorial that will allow _everyone_ to make up their _own_ minds if they want to _commit_ to `elm` or stick with the tools @andrewMacmurray listed above... 🔨 

> We are up to ***45 Votes*** on the Poll: https://twitter.com/nelsonic/status/795164336834617344 please ensure _everyone_ you know has voted! 🗳 

We will _know_ the _outcome_ of the `elm-lang` vote at the same time as the [`US Election`](https://www.youtube.com/watch?v=_sbT3_9dJY4) result; 
_hopefully_ both go the way of ❤️  not 😨 

> _still_ think [***`Riot`***](https://github.com/dwyl/hapi-riot) is _both_ _**conceptually**_ + _**practically**_ simpler than React: http://riotjs.com/compare/
**`And`** if people stick to the JS [***`goodparts`***](https://github.com/dwyl/goodparts) (_fewer language features focused on unambiguous code_) we can build _simple_ apps that are [***`easy`***](https://youtu.be/at3FPJaAwoY) for _others_ to ***maintain***. But, the more I look into the elm _ecosystem_ the more I am _convinced_ that it's a _viable_ option for us and the people/organisations we _serve_. 😉 

# tl;dr (_[when I were a lad](https://www.youtube.com/watch?v=Xe1a1wHxTyo)_...):

Rewind the clock back to **2011** when I was still writing code _full-time_ (_instead of reviewing other people's work and helping hire/help other team members ... `#nostalgia`_) We were writing (_relatively_) "_boring_" E-commerce web apps and it did not _feel_ particularly "_exciting_"... 

When I started using [**`CoffeeScript`**](http://coffeescript.org/) and could write white-space significant, compile-to-JS functional expressions, optimised (_"good parts"-only_) code with the brevity of Lisp I was _elated_! 🌈 🦄 

A _few_ people @Groupon were "_concerned_" that we wouldn't be able to "_hire_" good people if we picked an "_unknown_" language ... however using our internal "learning" docs we could teach people who had never written a `line` of `.coffee` in their lives `CoffeeScript` _**from-scratch**_ and get them up-and-running-and-productive in _**less than a day**_ on a _vast_ codebase written by _hundreds_ of devs... with the _right_ (_simple/focused_) language the ["_on-boarding_"](https://en.wikipedia.org/wiki/Onboarding) can be pretty smooth.
(_sadly the learning docs were closed source because they included lots of references to the actual codebase, otherwise `dwyl` would have a kickass `learn-coffeescript` repo...!!_) 

> The "_proof_" that CoffeeScript, Node.js & WebSockets is a "Winning Recipe" came from our friends at **Fog Creek**, who you _may_ know for their _insanely **successful**_ "`SAAS`" Product: **Trello**: https://trello.com/ see: https://www.fogcreek.com/blog/the-trello-tech-stack/ written in Jan 2012 _long_ before Node.js was "cool" or "trendy" ... 
PDF in case the blog post moves: [The-Trello-Tech-Stack.pdf](https://github.com/dwyl/technology-stack/files/577649/The-Trello-Tech-Stack.pdf)

Around the same time I watched **[`Bret Victor's - Inventing on Principle`](https://www.youtube.com/watch?v=PUv66718DII)** and it made me ["_believe in **magic**_"](https://www.youtube.com/watch?v=mDYNuD4CwlI) and I _regret_ not following my _instinct_ of quitting my "day job" to go do a PhD in Reactive coding based on @worrydream's work https://github.com/worrydream/Tangle ... but mortgage, bills-to-pay, yada yada ... 

If I had the _luxury_ of focussing exclusively on building tools & a _single **product**_ (_instead of client-projects which are good for keeping the lights on..._) the elm ecosystem is _almost_ exactly what I would _want_ to build:

+ Time-travelling (re-playable!) Reactive Debugger: http://debug.elm-lang.org/
+ Pure Functions with No-side-effects: https://en.wikipedia.org/wiki/Pure_function (_way simpler to understand than "OOP"_)
+ Immutable Data with Predictable State: https://github.com/dwyl/learn-react/issues/6

> Long-story-short, fast-forward to 2016, the Elm _ecosystem_ makes me feel like [**`this`**](https://youtu.be/ru0K8uYEZWw) `||` [***`this`***](https://youtu.be/k4OCkQPPyFk?t=3s)
**`And`** I think it would be _foolish_ to _waste_ the _enthusiasm_ a _few_ people have expressed for learning/using elm-lang. ❤️  ✅  🚀

**Comment by nelsonic** [comment]
The Twitter Poll `Final results` are in: https://github.com/dwyl/learn-elm/issues/10 ⚖ 
***`And`*** people are `elm-curious` so we need to write a _decent_ intro guide to help get them _excited_. 💃 

> **Note**: we are still _far_ from _answering the _original_ `Question`, but I'm _confident_ we will have an answer _soon_. ⌛️ 
> If you have _time_ please `help` by commenting on the issues or raising new ones you think of: https://github.com/dwyl/learn-elm/issues (_thanks_!) ❤️  ✅  🎉

**Comment by des-des** [comment]
I am still quite confused about why elm is the right tech for dwyl.

From looking at other dwyl projects, it seems like server-side rendering has been a priority. Assuming that this is still important I see two major problems.

Firstly Elm does not currently support server-side rendering. This is a problem that will eventually go away, but right now it does not seem easy.

The second problem seems much bigger to me. Elm seems like the wrong tech choice to do progressive enhancement. eg:
  1. We have a todo list, served as html.
  2. The user wants to sort a list of todos. If the client has js disabled then it will request a sorted todo list from the server.
  3. In the case where the client has js, we have two options.
     1. Make ajax request for json, create a sorted list of todos, render them. In this case we would need to re-write the view layer in elm. We will have a template for a todo item in elixir, and we will have a view function written in elm, this duplication feels inferior to old stack, where we could reuse handlebars templates.
     2. Shuffle the dom nodes. As far as I understand you would not mess directly with the dom in elm..

As far as I can see, elm + elixir are much nicer pieces of tech than node + js, but it seems like a weird choice for dwyl.. Basically elm is great for building client side apps, but right now, this is all that it is meant to do. Considering that dwyl has always favoured ssr over client side rendered apps, this decision leaves me totally confused.. :cry:

**Comment by nelsonic** [comment]
@des-des You're _right_ to be "_confused_". Until we have a canonical "Chat" example, it won't be clear how Phoenix and Elm can work together.

> Yes, elm does not support Server-Side-Rendering and that is _still_ our priority for ensuring that our Web Apps render and are served as fast as possible for Mobile clients.

What I will say to you and anyone `else` reading this is: _focus_ on Phoenix for the next couple of months and ***ignore*** the `Elm` in "PETE" for now. ~~PETE~~ PET! 👍 


Where Elm comes into play for us is when there is Frontend functionality such as what you built on "ART" and what @Conorc1000 / @roryc89 & @naazy are building for "OA" where the UX lends itself to in-browser validation, calculation, transparent server requests and re-rendering. 
In that respect Elm offers us an order-of-magnitude better "developer experience" (_nicer to write, reason about and maintain_) than the equivalent code in another "front-end framework".

Elm is going to play a _major_ role in the dwyl Products from about April 2017 onward ...
we are going to use Elm to _progressively_ enhance or "_Super Charge_" our server-side-rendered apps.
The code written for the client does not (should not) be rendered on the server. 
If/when Elm has server-side-rendering we will **_embrace_ it like a long-lost-loved-one**!
But until then we can use "PETE" with Phoenix views and use Elm for _specific_ enhancements.

Yes, that _might_ occasionally mean re-doing certain components in Elm that have already been built in Elixir Templates, but we can "_mount_" these Elm components in specific DOM elements and let Elm "handle" re-rendering _just_ that element. 

You won't need to _think_ about Elm _until_ you need to do "_complex_" interaction.
A Todo list does not require client-side JS or Elm for that matter. 
A Phoenix-based Todo List App with a reasonably low-latency internet connection will do round-trips to the server in _under_ **400ms** which is (_more than_) ["_fast enough_"](http://www.science.smith.edu/departments/neurosci/courses/bio330/pdf/94CurrBiolTovee.pdf)
We can get that interaction down to **50ms** with Elm but that is not the _focus_ of a "Beta" or "MVP" App ... the focus of an MVP is having _enough_ functionality to _Test/Validate_ the idea, no Front-end JS is required for this. 

I don't think Phoenix is going to be _everyone's_ "_cup of tea_" and that's **100% fine**.
But it's _definitely_ mine for _so many_ reasons.

The fact that the Phoenix and Elm communities have so much _overlap_
see/watch: https://www.youtube.com/watch?v=XJ9ckqCMiKk
gives me a _lot_ of hope that Elm will be server-side-rendered in Phoenix quite soon. 🔜 
`.then` it will be "Full Stack Elm" and the _whole_ world will shift to "PETE"!! 🎉 

For the people _on_ the Founders & Coders "Course" we need to keep their learning "_stable_",
there is a _lot_ to learn just with JS, Testing, Node, APIs, PostgreSQL, WebSockets, Heroku, Design, UX etc ... getting them to learn "PETE" would be seriously _overwhelming_.

I think we should get F&C people to learn Elm and use it with a Node.js "backend". 
This is a _very_ legitimate Elm use-case.
Once people have _finished_ the F&C "core program" we should _introduce_ them to Phoenix as a "Level Up" for their "Backend Skills".

To summarise, all DWYL apps can be built with _Just_ "PET". 
We will bring in Elm later. 👍

**Comment by iteles** [comment]
I think I mentioned this in the PR actually - there's definitely a need to clarify that Elm is for client-side enhancements and server-side rendering is still key for what we do @des-des, thanks for the reminder!

**Comment by des-des** [comment]
@nelsonic Thanks for the **awesome** answer + video! This makes sense. I agree that Elm is a much nicer + all round better experience than react so I am excited for the future! Also loving Pheonix so far.

**Comment by des-des** [comment]
Here is an example of elm / phoenix: https://github.com/CultivateHQ/seat_saver

**Comment by nelsonic** [comment]
@des-des the related blog post for that repo is: http://www.cultivatehq.com/posts/phoenix-elm-11/ 😉

**Comment by des-des** [comment]
Yeah looked at seat-saver its a rubbish example you could replace elm with js and would not have to make a single change to the server

**Comment by nelsonic** [comment]
@des-des yeah, I don't think it's meant to be a "_this is why Elm is 'better' than JS_" post/example ...
more of a "look what we can do without writing JS..."
I _Agree_ that we can do a better example. 😉

**Comment by des-des** [comment]
ops jus saw it was your link is what I was rubbishing, sorry for the negative language, meant to be rubbishing my link..

We know that we can include js as an asset, the elm here could just as well be angular .. Obvs since elm and phoenix both play nicely with sockets / real time there is some synergy here.

**Comment by nelsonic** [comment]
closing. 
see: https://github.com/dwyl/learn-elm-architecture-in-javascript and https://github.com/dwyl/learn-elm 🎉
